pytorch_lightning/core/optimizer.py

```diff
@@ -325,9 +329,29 @@ def _configure_schedulers_manual_opt(schedulers: list, monitor: Optional[str]) -
             lr_schedulers.append({**default_config, **scheduler})
         else:
             lr_schedulers.append({**default_config, "scheduler": scheduler})
+
     return lr_schedulers
 
 
+def _validate_scheduler_api(lr_schedulers: List[Dict[str, Any]], model: "pl.LightningModule") -> None:
+    from pytorch_lightning.trainer.connectors.checkpoint_connector import _SupportsStateDict
+
+    for scheduler_config in lr_schedulers:
+        scheduler = scheduler_config["scheduler"]
+        if not isinstance(scheduler, _SupportsStateDict):
+            raise ValueError(
+                f"The provided lr scheduler `{scheduler.__class__.__name__}` is invalid."
+                " It should have `state_dict` and `load_state_dict` methods defined."
+            )
+
+        if not isinstance(scheduler, LRSchedulerTypeTuple) and not is_overridden("lr_scheduler_step", model):
+            raise MisconfigurationException(
+                f"The provided lr scheduler `{scheduler.__class__.__name__}` doesn't follow the PyTorch LR Scheduler"
+                " Protocol. You should override the `LightningModule.lr_scheduler_step` hook with your own logic if"
+                " you are using a custom LR scheduler."
+            )
```

Comment:
couldn't find a way to check the argument types in case it has a step method. https://github.com/PyTorchLightning/pytorch-lightning/pull/10249#discussion_r778752315, so I thought maybe let users override this hook because we can't for sure say whether default implementation will work or not. is there any other way around?

Reply:
The other option is doing

```python
try:
    lightning_module.lr_scheduler_step()
except TypeError as e:
    if "step()" in str(e):
        raise MisconfigurationException(
            f"The provided lr scheduler `{scheduler.__class__.__name__}` doesn't follow the PyTorch LR Scheduler"
            " API. You should override the `LightningModule.lr_scheduler_step` hook with your own logic if"
            " you are using a custom LR scheduler."
        ) from e
    raise
```

Maybe also check `AttributeError: .* object has no attribute 'step'`

This would be similar to #10719

Reply:
yeah, thought of doing this.. but wasn't sure if it's clean...
why didn't timm just use PyTorch scheduler 😄 

will update.

Reply:
@carmocca after thinking a bit, I think keeping it as it is should be okay. The reason is if someone copies the exact implementation of let's stay StepLR without subclassing LR_Scheduler then it won't work anyway because it uses its superclass methods too to work. Raising an error during the actual step might be too late for cases where it happens at the epoch level. We can consider it like this: if someone's using native schedulers then Lightning will handle it by default else one has to override this hook. One case as you mentioned can be user copying the whole implementation including the superclass methods and not subclassing might be very very rare, and I don't think anyone might be doing that plus if someone chooses to do it anyways then we can't help since we don't do any such checks for Optimizer or any other PyTorch components in a similar case.

Reply:
Consider this case, adapted from your tests

```python
def test_user_is_forced_to_override_unnecesarily():

    class CustomScheduler:
        def __init__(self, optimizer):
            self.optimizer = optimizer

        def step(self):
            ...

        def state_dict(self):
            ...

        def load_state_dict(self, state_dict):
            ...

    class CustomBoringModel(BoringModel):
        def configure_optimizers(self):
            opt = torch.optim.SGD(self.parameters(), lr=1e-2)
            lr_scheduler = CustomScheduler(opt)
            return {"optimizer": opt, "lr_scheduler": lr_scheduler}

    model = CustomBoringModel()
    model.trainer = Trainer()
    _init_optimizers_and_lr_schedulers(model)
```

This fails unnecessarily. Forces you to override the method when the base method would work.

If you prefer not to do this for now, or leave this for a follow-up, I'm fine with that. Just wanted to point out that this would be a source of boilerplate for that small percentage of users.

Reply:
I see it as a very rare case. If there's a way to test it during init then it would be okay.. but my only concern regarding adding the exception during the actual step might be too late.